source/opt/licm_pass.cpp

```diff
@@ -0,0 +1,149 @@
+    // Ignore nested loops, as we will process them in order in ProcessLoop
+    if (loop.IsNested()) {
+      continue;
+    }
+    modified |= ProcessLoop(&loop, f);
+  }
+  return modified;
+}
+
+bool LICMPass::ProcessLoop(ir::Loop* loop, ir::Function* f) {
+  bool modified = false;
+
+  // Process all nested loops first
+  for (ir::Loop* nested_loop : *loop) {
+    modified |= ProcessLoop(nested_loop, f);
+  }
+
+  // Get the pre_header. If it doesn't exist, force it's creation
+  ir::BasicBlock* pre_header_bb = loop->GetPreHeaderBlock();
+  if (!pre_header_bb) {
+    modified = true;
+    pre_header_bb = loop->GetOrCreatePreHeaderBlock();
+  }
+  ir::BasicBlock* header_bb = loop->GetHeaderBlock();
+  ir::BasicBlock* merge_bb = loop->GetMergeBlock();
+  std::queue<ir::BasicBlock*> loop_bbs{};
+  modified |=
+      AnalyseAndHoistFromBB(loop, f, header_bb, merge_bb,
+                            pre_header_bb, loop->GetHeaderBlock(), &loop_bbs);
+
+  while (!loop_bbs.empty()) {
+    ir::BasicBlock* bb = loop_bbs.front();
+    loop_bbs.pop();
+    modified |= AnalyseAndHoistFromBB(loop, f, header_bb,
+                                      merge_bb, pre_header_bb, bb, &loop_bbs);
+  }
+
+  return modified;
+}
+
+bool LICMPass::AnalyseAndHoistFromBB(ir::Loop* loop, ir::Function* f,
+                                     ir::BasicBlock* header_bb,
+                                     ir::BasicBlock* merge_bb,
+                                     ir::BasicBlock* pre_header_bb,
+                                     ir::BasicBlock* bb,
+                                     std::queue<ir::BasicBlock*>* loop_bbs) {
+  bool modified = false;
+  if (!IsInsideNestedLoop(loop, f, bb)) {
+    bool hoisted = true;
+    while (hoisted) {
+      hoisted = false;
+      // When we perform a hoist it will invalidate this iterator, but we
+      // require the hoist so instructions later in the bb which depended on the
+      // hoisted instruction can now be hoisted. As such, we break iteration
+      // over the bb and start again with fresh iterators until no instructions
+      // are hoisted
+      for (ir::Instruction& inst : *bb) {
+        if (loop->IsInstructionInvariant(context(), &inst)) {
+          HoistInstruction(pre_header_bb, &inst);
+          modified = true;
+          hoisted = true;
+          break;
+        }
+      }
+    }
+  }
+
+  ir::CFG cfg(get_module());
+  opt::DominatorAnalysis* dom_analysis =
+      context()->GetDominatorAnalysis(f, cfg);
+  opt::DominatorTree& dom_tree = dom_analysis->GetDomTree();
+
+  for (opt::DominatorTreeNode* child_dom_tree_node :
+       *dom_tree.GetTreeNode(bb)) {
+    if (dom_analysis->Dominates(header_bb, child_dom_tree_node->bb_) &&
+        !dom_analysis->Dominates(merge_bb, child_dom_tree_node->bb_)) {
+      loop_bbs->push(child_dom_tree_node->bb_);
+    }
+  }
+
+  return modified;
+}
+
+bool LICMPass::IsInsideNestedLoop(ir::Loop* loop, ir::Function* f,
```

Comment:
The name implies something different that what you are checking.  What if `bb` is not in `loop` at all.  Then it is not in a loop that is nested in `loop`, but this function will return true.

A better name might be `IsImmediatelyContainInLoop` (seems a little long), or anything you can think of.